### PR TITLE
feat(node)!: Implement graceful shutdown

### PR DESCRIPTION
**Pull request by oblique**

### CODE REVIEW

File: node/src/p2p/header_ex/client.rs [hunk 1]
```diff
@@ -143,45 +187,68 @@ where
...
+                    }
                 }
-            }
 
-            // Otherwise return the header with the maximum height
-            let resp = resps.into_iter().next().expect("no reposnes");
-            respond_to.maybe_send_ok(vec![resp]);
-        });
+                // Otherwise return the header with the maximum height
+                let resp = resps.into_iter().next().expect("no reposnes");
```
Comment (fl0rek) [inline node/src/p2p/header_ex/client.rs:247]:
```suggestion
                let resp = resps.into_iter().next().expect("no responses");
```

File: node/src/utils/token.rs [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+            token: CancellationToken::new(),
+        }
+    }
+
+    /// Trigger the event
+    pub(crate) fn trigger(&self) {
+        self.token.cancel();
+    }
+
+    /// Returns a guard that will trigger the event on drop.
```
Comment (fl0rek) [inline node/src/utils/token.rs:22]:
```suggestion
    /// Returns a guard that will trigger the token on drop.
```

File: node/src/utils/token.rs [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+        }
+    }
+
+    /// Returns if event is triggered or not.
+    #[allow(dead_code)]
+    pub(crate) fn is_triggered(&self) -> bool {
+        self.token.is_cancelled()
+    }
+
+    /// Returns when the event is triggered. If it was triggered before,
```
Comment (fl0rek) [inline node/src/utils/token.rs:35]:
```suggestion
    /// Returns when the token is triggered. If it was triggered before,
```
I think it's touch clearer when talking about just the token, without involving potential events user might stack on top of it.

**Review by zvolin — APPROVED** [review]
nice, looks solid :mechanical_arm:

**Review by fl0rek — APPROVED** [review]
LGTM!

**Pull request merged**